object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java

```diff
@@ -0,0 +1,515 @@
+   * otherwise.
+   */
+  private static LocalVarWithTree getAssignmentTreeOfVar(
+      Set<LocalVarWithTree> defs, LocalVariableNode node) {
+    return defs.stream()
+        .filter(assign -> assign.localVar.getElement().equals(node.getElement()))
+        .findAny()
+        .orElse(null);
+  }
+
+  /** checks if the variable has been declared in a try-with-resources header */
+  private static boolean isTryWithResourcesVariable(LocalVariableNode lhs) {
+    Tree tree = lhs.getTree();
+    return tree != null
+        && TreeUtils.elementFromTree(tree).getKind().equals(ElementKind.RESOURCE_VARIABLE);
+  }
+
+  /**
+   * Creates the appropriate @CalledMethods annotation that corresponds to the @MustCall annotation
+   * declared on the class type of {@code localVarWithTree.first}. Then, it gets @CalledMethod
+   * annotation of {@code localVarWithTree.first} to do a subtyping check and reports an error if
+   * the check fails.
+   */
+  private void checkMustCall(
+      LocalVarWithTree localVarWithTree, CFStore store, String outOfScopeReason) {
+    List<String> mustCallValue = typeFactory.getMustCallValue(localVarWithTree.tree);
+    // optimization: if there are no must-call methods, we do not need to perform the check
+    if (mustCallValue.isEmpty()) {
+      return;
+    }
+    AnnotationMirror cmAnnoForMustCallMethods =
+        typeFactory.createCalledMethods(mustCallValue.toArray(new String[0]));
+
+    AnnotationMirror cmAnno;
+
+    // sometimes the store is null!  this looks like a bug in checker dataflow.
+    // TODO track down and report the root-cause bug
+    CFValue lhsCFValue = store != null ? store.getValue(localVarWithTree.localVar) : null;
+    if (lhsCFValue != null) { // When store contains the lhs
+      cmAnno =
+          lhsCFValue.getAnnotations().stream()
+              .filter(anno -> AnnotationUtils.areSameByClass(anno, CalledMethods.class))
+              .findAny()
+              .orElse(typeFactory.top);
+    } else {
+      cmAnno =
+          typeFactory
+              .getAnnotatedType(localVarWithTree.localVar.getElement())
+              .getAnnotationInHierarchy(typeFactory.top);
+    }
+
+    if (!typeFactory.getQualifierHierarchy().isSubtype(cmAnno, cmAnnoForMustCallMethods)) {
+      if (!reportedMustCallErrors.contains(localVarWithTree)) {
+        reportedMustCallErrors.add(localVarWithTree);
+
+        checker.reportError(
+            localVarWithTree.tree,
+            "required.method.not.called",
+            formatMissingMustCallMethods(mustCallValue),
+            localVarWithTree.localVar.getType().toString(),
+            outOfScopeReason);
+      }
+    }
+  }
+
+  /**
+   * Is {@code exceptionClassName} an exception type we are ignoring, to avoid excessive false
+   * positives? For now we ignore {@code java.lang.Throwable} and {@code NullPointerException}
+   */
+  private static boolean isIgnoredExceptionType(Name exceptionClassName) {
```

Comment:
Can we add a Signature Checker annotation to `exceptionClassName` to indicate the kind of name we're expecting here? I find those annotations very useful just as documentation, even if we aren't running the signature checker (yet?).

Reply:
caba8d5